SQL: Failure of query containing two or more "UNION ALL"

## Details

This query currently produces a panic dump in SuperDB.

```
$ super -c "
SELECT a FROM t1.sup
UNION ALL
SELECT a FROM t2.sup
UNION ALL
SELECT a from t3.sup;"

panic: semSQLOp: unknown op: &ast.Union{Kind:"Union", Distinct:false, Left:(*ast.Select)(0xc0004a4c80), Right:(*ast.Select)(0xc0004a4fa0), Loc:ast.Loc{First:32, Last:82}}

goroutine 1 [running]:
github.com/brimdata/super/compiler/semantic.(*analyzer).semSQLOp(0xc0004ed850?, {0xe72f6d0?, 0xc000813310?}, {0x0?, 0xc0003c6340?, 0x3?})
	/Users/phil/work/super/compiler/semantic/sql.go:504 +0x6e5
github.com/brimdata/super/compiler/semantic.(*analyzer).semOp(0xc0004ed850, {0xe72f6d0?, 0xc000813950}, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/semantic/op.go:937 +0x204b
github.com/brimdata/super/compiler/semantic.(*analyzer).semSeq(0xc0004ed850, {0xc0003a7a50, 0x1, 0x1})
	/Users/phil/work/super/compiler/semantic/op.go:33 +0x51
github.com/brimdata/super/compiler/semantic.Analyze({0xe735980, 0xc0003c8660}, 0xc0003a1d80, 0xc0003a1e00, 0x0)
	/Users/phil/work/super/compiler/semantic/analyzer.go:25 +0x325
github.com/brimdata/super/compiler.Analyze(...)
	/Users/phil/work/super/compiler/package.go:27
github.com/brimdata/super/compiler.CompileWithAST(0xc0003c8660, 0xc8f8405?, 0xc0003a1e00, 0x1, 0xc, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/package.go:74 +0x65
github.com/brimdata/super/compiler.(*compiler).NewQuery(0xe735a28?, 0xc0003c0dc0?, 0xc0004ed9a8?, {0x0?, 0x8?, 0xe4e6980?}, 0x1?)
	/Users/phil/work/super/compiler/compiler.go:42 +0x45
github.com/brimdata/super/runtime.CompileQuery({0xe735a28?, 0xc0003c0dc0?}, 0xc00043eea0, {0xe724238, 0xc00019e190}, 0xc0003a1d80, {0x0, 0x0, 0x0})
	/Users/phil/work/super/runtime/compiler.go:34 +0x10f
github.com/brimdata/super/cmd/super/root.(*Command).Run(0xc000128248, {0xc00019a6e0, 0x0, 0x0})
	/Users/phil/work/super/cmd/super/root/command.go:158 +0x3c5
github.com/brimdata/super/pkg/charm.path.run({0xc00012a7c0, 0x1, 0x1}, {0xc00019a6e0, 0x0, 0x0?})
	/Users/phil/work/super/pkg/charm/path.go:11 +0x7b
github.com/brimdata/super/pkg/charm.(*Spec).Exec(0xf875c60, {0xc00019a6d0, 0x2, 0x2})
	/Users/phil/work/super/pkg/charm/charm.go:74 +0x1fa
main.main()
	/Users/phil/work/super/cmd/super/main.go:39 +0x5b
```

## Details

Repro is with super commit dfcf85fcf. This was found via a query from [a sqllogictest](https://github.com/duckdb/duckdb/blob/1de84b63461662bb6b56f20cea5a6698fd759622/test/sqlite/select4.test_slow#L3495-L3531).

Here's the query running successfully in Postgres:

```
$ psql postgres
psql (17.5 (Homebrew))
Type "help" for help.

postgres=# CREATE TABLE t1(a INTEGER);
CREATE TABLE
postgres=# CREATE TABLE t2(a INTEGER);
CREATE TABLE
postgres=# CREATE TABLE t3(a INTEGER);
CREATE TABLE
postgres=# INSERT INTO t1 VALUES(1);
INSERT 0 1
postgres=# INSERT INTO t2 VALUES(2);
INSERT 0 1
postgres=# INSERT INTO t3 VALUES(3);
INSERT 0 1

postgres=# SELECT a from t1
UNION ALL
SELECT a from t2
UNION ALL
SELECT a from t3;

 a 
---
 1
 2
 3
(3 rows)
```

SuperDB works ok as long as there's only a single `UNION ALL`, e.g.,

```
$ cat t1.sup 
{a:1}

$ cat t2.sup 
{a:2}

$ cat t3.sup 
{a:3}

$ super -version

$ super -c "
SELECT a FROM t1.sup
UNION ALL
SELECT a FROM t2.sup;"

{a:2}
{a:1}
```

--- a/compiler/semantic/op.go
+++ b/compiler/semantic/op.go
@@ -508,7 +508,7 @@ func (a *analyzer) semDebugOp(o *ast.Debug, mainAst ast.Seq, in dag.Seq) dag.Seq
 // with either an aggregate or filter op based on the function's name.
 func (a *analyzer) semOp(o ast.Op, seq dag.Seq) dag.Seq {
 	switch o := o.(type) {
-	case *ast.Select, *ast.SQLLimitOffset, *ast.OrderBy, *ast.SQLPipe, *ast.With, *ast.SQLValues:
+	case *ast.Select, *ast.SQLLimitOffset, *ast.OrderBy, *ast.SQLPipe, *ast.Union, *ast.With, *ast.SQLValues:
 		seq, sch := a.semSQLOp(o, seq)
 		seq, _ = derefSchema(sch, seq)
 		return seq
@@ -866,19 +866,6 @@ func (a *analyzer) semOp(o ast.Op, seq dag.Seq) dag.Seq {
 			},
 		})
 		return append(seq, dag.NewValues(&dag.This{Kind: "This", Path: field.Path{"sample"}}))
-	case *ast.Union:
-		if o.Distinct {
-			a.error(o, errors.New("UNION DISTINCT not currently supported"))
-			return append(seq, badOp())
-		}
-		left, leftSch := a.semSQLOp(o.Left, seq)
-		left, _ = derefSchema(leftSch, left)
-		right, rightSch := a.semSQLOp(o.Right, seq)
-		right, _ = derefSchema(rightSch, right)
-		return dag.Seq{
-			&dag.Fork{Kind: "Fork", Paths: []dag.Seq{left, right}},
-			&dag.Combine{Kind: "Combine"},
-		}
 	case *ast.Assert:
 		cond := a.semExpr(o.Expr)
 		// 'assert EXPR' is equivalent to

--- a/compiler/semantic/sql.go
+++ b/compiler/semantic/sql.go
@@ -483,6 +483,19 @@ func (a *analyzer) semSQLOp(op ast.Op, seq dag.Seq) (dag.Seq, schema) {
 			out = append(out, &dag.Head{Kind: "Head", Count: a.evalPositiveInteger(op.Limit)})
 		}
 		return out, schema
+	case *ast.Union:
+		if op.Distinct {
+			a.error(op, errors.New("UNION DISTINCT not currently supported"))
+		}
+		left, leftSch := a.semSQLOp(op.Left, seq)
+		left, _ = derefSchema(leftSch, left)
+		right, rightSch := a.semSQLOp(op.Right, seq)
+		right, _ = derefSchema(rightSch, right)
+		return dag.Seq{
+			&dag.Fork{Kind: "Fork", Paths: []dag.Seq{left, right}},
+			&dag.Combine{Kind: "Combine"},
+		}, &dynamicSchema{}
+
 	case *ast.With:
 		if op.Recursive {
 			a.error(op, errors.New("recursive WITH queries not currently supported"))

--- a/compiler/ztests/sql/union-all.yaml
+++ b/compiler/ztests/sql/union-all.yaml
@@ -1,5 +1,5 @@
 script: |
-  super -s -c 'SELECT * FROM a.sup UNION ALL SELECT * FROM b.sup |> sort'
+  super -s -c 'SELECT * FROM a.sup UNION ALL SELECT * FROM b.sup UNION ALL SELECT * from c.sup |> sort x'
 
 vector: true
 
@@ -11,6 +11,9 @@ inputs:
     data: |
       {x:1,y:2}
       {y:3,x:4}
+  - name: c.sup
+    data: |
+      {x:5}
 
 outputs:
   - name: stdout
@@ -18,3 +21,4 @@ outputs:
       {x:1,y:2}
       {x:1,y:2}
       {y:3,x:4}
+      {x:5}

